packages/phpstan-rules/src/Rules/CheckConstantExpressionDefinedInConstructOrSetupRule.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Symplify\PHPStanRules\Rules;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\Assign;
+use PhpParser\Node\Expr\BinaryOp\Concat;
+use PhpParser\Node\Expr\ClassConstFetch;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\Variable;
+use PhpParser\Node\Scalar\MagicConst;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\NodeFinder;
+use PHPStan\Analyser\Scope;
+use Symplify\PHPStanRules\ValueObject\PHPStanAttributeKey;
+
+/**
+ * @see \Symplify\PHPStanRules\Tests\Rules\CheckConstantExpressionDefinedInConstructOrSetupRule\CheckConstantExpressionDefinedInConstructOrSetupRuleTest
+ */
+final class CheckConstantExpressionDefinedInConstructOrSetupRule extends AbstractSymplifyRule
+{
+    /**
+     * @var string
+     */
+    public const ERROR_MESSAGE = 'Constant expression should only defined in __construct() or setUp()';
+
+    /**
+     * @var NodeFinder
+     */
+    private $nodeFinder;
+
+    public function __construct(NodeFinder $nodeFinder)
+    {
+        $this->nodeFinder = $nodeFinder;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [Assign::class];
+    }
+
+    /**
+     * @param Assign $node
+     * @return string[]
+     */
+    public function process(Node $node, Scope $scope): array
+    {
+        $classMethod = $this->resolveCurrentClassMethod($node);
+        if ($classMethod === null) {
+            return [];
+        }
+
+        $parent = $node->getAttribute(PHPStanAttributeKey::PARENT)
+            ->getAttribute(PHPStanAttributeKey::PARENT);
+        if (! $parent instanceof ClassMethod || $this->isFoundInNextStatement($node)) {
+            return [];
+        }
+
+        if (in_array(strtolower((string) $classMethod->name), ['__construct', 'setup'], true)) {
+            return [];
+        }
+
+        if ($node->expr instanceof Concat && $node->expr->left instanceof MagicConst) {
+            if ($node->expr->right instanceof MethodCall) {
+                return [];
+            }
+
+            return [self::ERROR_MESSAGE];
+        }
+
+        if (! $node->expr instanceof ClassConstFetch) {
+            return [];
+        }
```

Comment:
Maybe, the evaluation of constant expression could be handled with:
`PhpParser\ConstExprEvaluator` (I haven't tested it, just remembered it might help)

Reply:
Has the service been useful at all?

Reply:
that seems for check const in many cases, while above, I only check instance of ClassConstFetch, not sure how useful that for it

Reply:
I see. We need to check for any constant expressoin, not just ClassConstFetch.
All of these should be evaluated as constnat expressions:

```php
$staticValue = __DIR__ . '/integer';

$staticValue = __DIR__ . 100 . 'value' . DIRETORY_SEPARATOR;

$staticValue = __DIR__ . 100 . 'value' . DIRETORY_SEPARATOR;
```

And these should be skipped

```php
$staticValue = __DIR__ . self::getValue();

```

Reply:
the MagicConst assignment already handled before its check, is current implementation already correct? There are already fixture for that

Reply:
Let's give a try. I'll merge it and try to break it :) 

Thank you

Reply:
It's broken. This test shoud pass:
https://github.com/symplify/symplify/pull/2458

Reply:
I will check